plugins/RemovableDriveOutputDevice/DriveFilter.py

```diff
@@ -0,0 +1,118 @@
+# Copyright (c) 2022 Ultimaker B.V.
+# Uranium is released under the terms of the LGPLv3 or higher.
+
+import json
+import re
+from json.decoder import JSONDecodeError
+from typing import Dict, List
+from cura.CuraApplication import CuraApplication
+from UM.Logger import Logger
+
+
+class DriveFilter:
+    def __init__(self):
+        """Creates the filter.
+        """
+
+        # Set the initial filter steps to be empty.
+        self.filterSteps: List[Dict[str, str]] = []
+
+        # Set up the preferences.
+        # A None check is done for the unit tests, which don't test the application.
+        curaApplication = CuraApplication.getInstance()
+        if curaApplication is not None:
+            preferences = curaApplication.getPreferences()
+            preferences.addPreference("removable_drive_output_device/filter_steps", "[]")
+            preferences.preferenceChanged.connect(self._onPreferencesChanged)
+            self.reloadFromPreferences()
+
+    def reload(self, filterSteps: List[Dict[str, str]]) -> None:
+        """Reloads the filter.
+
+        :param filterSteps: Steps in the filter to apply.
+        """
+
+        self.filterSteps = filterSteps
+
+    def reloadFromPreferences(self) -> None:
+        """Reloads the drive filter from the Cura preferences.
+        """
+
+        preferences = CuraApplication.getInstance().getPreferences()
+        try:
+            # Load the filter steps.
+            self.filterSteps = json.loads(preferences.getValue("removable_drive_output_device/filter_steps"))
+            Logger.info("Loaded drive filter steps from preferences.")
+        except JSONDecodeError:
+            # Display an error that the stored filter steps aren't valid JSON.
+            Logger.error("Drive filter steps are invalid JSON. Unable to load from preferences.")
+
+    def _onPreferencesChanged(self, name: str) -> None:
+        """Handles a preference being changed.
+
+        :param name: Name of the preference that changed.
+        """
+
+        # Reload the filter steps if the filter steps preference changed.
+        if name != "removable_drive_output_device/filter_steps":
+            return
+        self.reloadFromPreferences()
+
+    def passesFilter(self, string: str) -> bool:
+        """Returns if a string can pass the filter.
+        A string is considered passing the filter if:
+        1. It passes at least 1 whitelist condition.
+        2. It passes every blacklist condition.
+
+        :param string: String to test.
+        :return: Whether the string passes the filter.
+        """
```

Comment:
It seems that this whole function can be simplified to a single regex, even. Something like this (untested):

    blacklists = "&".join([("^(" + step["pattern"] + ")") for step in self.filterSteps if step["type"] == "blacklistregex"])
    whitelists = "|".join([step["pattern"] for step in self.filterSteps if step["type"] == "whitelistregex"])
    return re.search("^(" + blacklists + ")&" + whitelists, string) is not None

That kind of goes to show that maybe it's sufficient to provide a single whitelist, or a single whitelist and a single blacklist. I think that would make it a lot easier for people to use this feature, since they don't have to type out a whole JSON structure which is error-prone if you need to just have a simple filter.

Reply:
I went with the original design to make adding new filter types easier to add. The new design is much more compact but I am concerned about what happens if a new user wants to use Cura at the scale of a makerspace and the whitelist/blacklist regex isn't suitable (no clue what though).

As-is, both ideas have some trade-offs without a UI to act as an optional way to edit it: the current implementation requires the user to understand JSON while the potentially simpler implementation requires the user to understand regex. Given the option between the two, I would assume the Lab Director would prefer editing JSON over editing regex because... it's regex. I could get an answer, but either way, if changes were requested, I wouldn't be making them since I graduated, don't use Cura personally, and don't have a build environment set up anymore.

Reply:
Any update on this? Without a response covering if this is fine or if the design won't be accepted without changes, I can't make said changes. I'd prefer to have all of this resolved before The Construct @ RIT resumes normal operations with the upcoming school year starting in a few weeks.

Reply:
Again, any updates on this? With PR 777 merged in Uranium, this is the last outstanding pull request I have from school before I can stop using this GitHub account.